### PR TITLE
fix: use id as tiebreaker in cursor-based pagination to handle duplicate timestamps

### LINKED ISSUE #9357: [BUG]: pagination breaks when there are multiple records with the same created time stamp
**Describe the bug**
If you are listing an entity (i.e flow_runs) and you have 50 runs with the same created time stamp (when you bulk retry for example), and you have the page size set to 10, and you are only viewing the retried runs, it will show only the first 10 runs then when you click next, no runs will be listed, the reason is the paginator culls the extra results, and the next cursor will ask for a created timestamp that is greater than the runs' created timestamp, so it will return an empty list.

### PR DESCRIPTION
**Pull request by ragini-pandey**
## What does this PR do?

Fixes cursor-based pagination breaking when multiple records share the same `created` timestamp. This happens when bulk-retrying flow runs — all retried runs get the same timestamp, causing the paginator to skip records after the first page.

### Explain How the Feature Works

The paginator now uses the entity `id` as a secondary sort key (tiebreaker) alongside the `created` timestamp. The cursor encodes both `created` and `id`, and the WHERE clause becomes:

```sql
(created < cursor_created)
OR (created = cursor_created AND id < cursor_id)
```

This ensures that even when many records share the same second-truncated timestamp, pagination proceeds correctly through all of them.

**Changes in `paginator.ts`:**

1. **Cursor encoding** — now includes `id` alongside `created` in the cursor payload
2. **Cursor query** — uses a composite condition with `id` as tiebreaker when timestamps match
3. **Sort order** — adds `id` as secondary ORDER BY to guarantee deterministic ordering
4. **Type detection** — fixed `getEntityPropertyType` to handle constructor types (`String`, `Number`, `Date`) used in EntitySchema column definitions

The fix is backward-compatible — old cursors without `id` gracefully fall back to the original timestamp-only comparison.

### Tests

Added 18 unit tests in `packages/server/api/test/unit/app/helper/pagination/paginator.test.ts`:

- Cursor encode/decode includes both `created` and `id`
- `ORDER BY` includes `id` as secondary sort key in both ASC and DESC
- Pagination works correctly with duplicate timestamps
- Backward compatibility with old cursors (created-only, no `id`)
- `getEntityPropertyType` handles constructor types (`String`, `Number`, `Date`)
- Composite orderBy is unaffected by changes
- Unlimited pagination behavior
- Basic pagination edge cases: empty results, within-limit, hasMore

All 55 unit tests pass (including the 18 new ones).

### Performance Assessment

No performance regression expected:

- The secondary `ORDER BY id` uses the primary key index — zero additional index cost
- The composite WHERE clause `(created < X) OR (created = X AND id < Y)` is a standard tiebreaker pattern that PostgreSQL optimizes using existing indexes
- Cursor payload grows by ~30 bytes (base64-encoded id) — negligible HTTP overhead

### Public API Impact

None. The cursor is an opaque base64 string in all public documentation and OpenAPI spec. Consumers pass `next`/`previous` values verbatim. The internal cursor format change is invisible to API users.

### Relevant User Scenarios

- Bulk retrying flow runs creates many records with identical timestamps, making it impossible to paginate beyond the first page
- Any scenario where entities are created in rapid succession within the same second

Fixes #9357

### CODE REVIEW

File: packages/server/api/src/app/helper/pagination/paginator.ts [hunk 1]
```diff
@@ -185,9 +185,30 @@ export default class Paginator<Entity extends ObjectLiteral> {
...
+                    { cursor_id: cursors['id'] },
+                )
+            }))
+        }
+        else {
+            where.orWhere(
+                `DATE_TRUNC('second', ${this.alias}.${PAGINATION_KEY}) ${operator} DATE_TRUNC('second', :${PAGINATION_KEY}::timestamp)`,
+                paginationKeyParam,
+            )
+        }
```
[the remark this answers is not in the document]
Comment (ragini-pandey) [reply packages/server/api/src/app/helper/pagination/paginator.ts]: Good catch — addressed in c3fb522. Hoisted the common orWhere call to run unconditionally; the id tiebreaker bracket is now only conditionally appended.

File: packages/server/api/src/app/helper/pagination/paginator.ts [hunk 2]
```diff
@@ -262,6 +283,7 @@ export default class Paginator<Entity extends ObjectLiteral> {
 
         const orderByCondition: Record<string, Order> = {}
         orderByCondition[`${this.alias}.${this.orderBy}`] = order
+        orderByCondition[`${this.alias}.id`] = order
```
[the remark this answers is not in the document]
Comment (ragini-pandey) [reply packages/server/api/src/app/helper/pagination/paginator.ts:281]: The primary key index on id is already well-optimized for tiebreaking. PostgreSQL can efficiently combine the existing (projectId, created) index with the PK index for the secondary sort. A dedicated composite (created, id) index would add write overhead on every insert across all entities. If profiling on high-traffic tables like flow_run shows regression, a targeted composite index can be added as a follow-up — but in typical usage, the PK tiebreaker should have negligible impact since most pages span distinct timestamps.

**Comment by abuaboud** [comment]
Hey — thanks so much for taking the time to open this! 🙏

We've **temporarily paused unsolicited pull requests** from outside the core team. This isn't about you or the quality of your change — an open PR is a promise from us to review it carefully, and right now the volume (a lot of it AI-generated) means we can't keep that promise for every PR without it becoming unfair to contributors who wait weeks for a reply. So instead of leaving PRs hanging, we close them and point you to a faster path.

**Already working with us?** If you're an existing customer and we've already discussed this change over Slack, ping us there and we'll add the `keep-open` label and reopen this PR for review.

You can still build and publish your own [piece](https://www.activepieces.com/docs/build-pieces/building-pieces/start-building) — just outside this repo, as your own package rather than a PR here. See [CONTRIBUTING.md](https://github.com/activepieces/activepieces/blob/main/CONTRIBUTING.md#pull-requests) for the full policy. Thanks again for caring about the project. 💜
